File: Blind75/98. Validate Binary Search Tree/98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */



class Solution {
public:
    // we could either inorder traverse the tree or use recursive to find out 
    bool isValidBST(TreeNode* root) {
        return helper(root,NULL,NULL);
    }
    
    bool helper(TreeNode* node, TreeNode* minVal, TreeNode* maxVal){
        if(node==NULL) return true;
        if(minVal!=NULL && node->val <= minVal->val) return false;
        if(maxVal!=NULL && node->val >= maxVal->val) return false;
        
        
        return helper(node->left,minVal,node)&&helper(node->right,node,maxVal);
        
        
        
    }
    
}; 

File: Blind75/98. Validate Binary Search Tree/98. Validate Binary Search Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        return helper(root, null, null);
    }

    bool helper(TreeNode node, TreeNode largest, TreeNode smallest){
        //base case
        if(node==null) return true;
        if(largest!=null && node.val >= largest.val) return false;
        if(smallest!=null && node.val <= smallest.val) return false;
        return helper(node.left, node, smallest) && helper(node.right,largest, node);

    }
}
